### DISCUSSION TITLE
Whats your opinion on bun.js

### DISCUSSION THREAD

**Discussion by mtsewrs** [uNetworking/uWebSockets#1466] [General]
What do think about bun? (Server based on uws)

https://bun.sh/

**Comment by FlorianBELLAZOUZ** [comment]
If we trust the benchmark bun have a faster FFI than nodejs (~x2.4) .

![](https://i.imgur.com/n3rS2bI_d.webp?maxwidth=760&fidelity=grand)

The project start just 1 year ago.
It's a totally new runtime base on JavaScriptCore.

So It's sound promising, but i'm afraid about the stability in real life.

What do you think @alexhultman, did you work with them for the uws integration ?

**Reply by prettydiff** [reply]
Where did the Node numbers come from?  I ask because Node does not provide an official API for web sockets, so web socket handling comes from user land (NPM packages).  Most of the popular Node web socket libraries I have seen implement web sockets over HTTP, so there is some opportunity to boost performance when implemented directly as a TCP/TLS socket using Node's Net and TLS libraries.

**Reply by uNetworkingAB** [reply]
It literally says right under "view source".

**Comment by FlorianBELLAZOUZ** [comment]
My quick Benchmark :

uws.js
```
const Uws = require('uWebsockets.js')

const app = Uws.SSLApp(
  { key_file_name:'/Users/Florian/lab/bun/key'
  , cert_file_name:'/Users/Florian/lab/bun/cert'})

app.get('*',res=>res.end('hi! :)'))

app.listen(8080,()=>{}) 
```

bun.ts
```
Bun.serve(
  { fetch:req=>new Response('hi! :)')
  , keyFile: './key'
  , certFile: './cert'
  , port:8080 })
```

**uws.js with ssl:**
florian@4444 ~ % wrk https://test.kaaarot.com:8080 
Running 10s test @ https://test.kaaarot.com:8080
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    74.12us  108.82us  11.36ms   99.87%
    Req/Sec    63.08k     1.96k   65.53k    98.02%
  1266546 requests in 10.10s, 73.68MB read
Requests/sec: **125395.76**
Transfer/sec:      7.29MB

**bun.js with ssl:**
florian@4444 ~ % wrk https://test.kaaarot.com:8080
Running 10s test @ https://test.kaaarot.com:8080
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    80.91us  144.57us  13.30ms   99.87%
    Req/Sec    59.47k     2.14k   63.65k    97.03%
  1194377 requests in 10.10s, 102.51MB read
Requests/sec: **118256.28**
Transfer/sec:     10.15MB

**uws.js without ssl:**
florian@4444 ~ % wrk http://test.kaaarot.com:8080
Running 10s test @ http://test.kaaarot.com:8080
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    59.79us   22.15us   1.48ms   75.24%
    Req/Sec    76.11k     1.64k   78.27k    91.58%
  1526538 requests in 10.10s, 88.81MB read
Requests/sec: **151146.51**
Transfer/sec:      8.79MB

**bun.js without ssl:**
florian@4444 ~ % wrk http://test.kaaarot.com:8080
Running 10s test @ http://test.kaaarot.com:8080
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    63.72us   25.52us   1.52ms   77.31%
    Req/Sec    72.15k     1.68k   77.05k    92.57%
  1449083 requests in 10.10s, 124.38MB read
Requests/sec: **143476.43**
Transfer/sec:     12.31MB

**Comment by mtsewrs** [comment]
This should've probably been posted in uWebSockets.js

**Comment by ghost** [comment]
Looks pretty slick I guess, but this will be an uphill battle for years to come

**Reply by FlorianBELLAZOUZ** [reply]
Yeah, and the project scope is hudge and the guy is alone.

So he didn't contact you for the uws integration i guess ?

**Reply by ghost** [reply]
I don't think he did

**Comment by ghost** [comment]
It's pretty obvious Deno is nothing but a cash grab profiting from Node.js's success - they did strict CLA from start, they raised millions of silicon valley funding and they sell their Deno deploy as their paid product (all they missed doing with nodejs while it blew up).

Good for Ryan Dahl, but for all who want something actually fast and efficient it sucks just like nodejs.

So whatever this bun thing is, if they can show these perf numbers which almost reach uWS.js levels I would say that's a very good foundation and I would like to see Deno/nodejs raise their perf to try and match it

**Comment by ghost** [comment]
Honestly, without overhyping this, bun is pretty much the ideal outcome of uWS - he seems to reimplement nodejs, but without being slow as a snail. That's the, in my opinion, only logical conclusion to draw based on uWS's perf numbers -

I would have done this myself if I had the time. The fact he's using nodejs compatible interfaces, but implementing them faster is going to help him get users very fast and is very clever.

What people who cannot put two and two together miss, is that uWS only shows one thing: that nodejs is a trainwreck in terms of performance. People can't use their brain and think "well if uWS is so much faster, that must mean nodejs is slow as shit". This guy made the connection and took the message. Whatever he does, it will be at least 5x better than nodejs is currently doing.

So yeah in essence, I think this is very interesting and very much needed.

**Comment by ghost** [comment]
Tried it, it rocks. Deno being "written with the fastest server tech available" sure performs like shit in comparison:

alexhultman@Alexs-MacBook-Air uSockets % ./http_load_test 40 localhost 3000
Running benchmark now...
Req/sec: 169417.250000
Req/sec: 169723.000000
Req/sec: 170540.750000
^C
alexhultman@Alexs-MacBook-Air uSockets % ./http_load_test 40 localhost 8080
Running benchmark now...
Req/sec: 80274.500000
Req/sec: 81097.000000
Req/sec: 80658.500000
Req/sec: 81197.750000

**Reply by NightScript370** [reply]
They probably mean at the time of writing.

**Reply by ghost** [reply]
That doesn't make it any less hocus pocus, pulled from ass made up nonsense. It's not true today and it never were

**Comment by ghost** [comment]
I don't think it is based on uWS, looks more like he experimented with it but ultimately made his own implementation. Still, it's way faster than Deno and Node.js and that's what counts.

Problem is funding and being burned out - this is a massive project covering a lot of area, and sooner or later you will get burned out and/or get a paying job instead, because -

Open source is kind of, unsustainable, since there's little money and the time you spend on it can be waaaaaay more personally rewarding by just getting a good job instead.

So question is how long this project can survive - the code looks pretty experimental and will need much polishing. The guy is going to be declared a rock star for sure but then what? His own best interest will be to just knock on Google's door and get himself a high paying job instead.

When you get 20k GitHub stars in a week you overload dopamine receptors and then when everything calms down you stand there with a massive heap of unfinished work and no payment, depression sets in.

So yeah, I love it but realistically it will be uphill from now

**Reply by Jarred-Sumner** [reply]
@alexhultman `Bun.serve` is based on uWebSockets and Bun's builtin `WebSocket` class uses uSockets (haven't implemented WebSocket server support yet). I made a few small tweaks, but mostly it uses uWebSockets. I think most of the performance difference between `uWS.js` and `Bun.serve` is due to Web APIs - `Request`, `Response`, `Headers`, etc have to be created which is unfortunate (it does pool `Response` objects at least).

`Bun.serve` mostly just wraps uWebSockets - you can [see the code](https://github.com/Jarred-Sumner/bun/blob/abd9d8bcab2eb3bb33c0f1df73e71ecf188ef8f1/src/bun.js/api/server.zig) 

[`WebSocket` client](https://github.com/Jarred-Sumner/bun/blob/f05428e4f3fd801242208e37def1ad6d5bad0fde/src/http/websocket_http_client.zig)

I'm thinking about using usockets for the builtin sockets library

**Reply by ghost** [reply]
Hmmm, interesting! Did you wrap the C++ in C then C in Zig? We have a PR to bring an official C interface of uws here (WIP). I'm biased but using uSockets sounds sane, since it already is C and gives you TCP, TLS, UDP and (soon) QUIC/HTTP3 with good perf.

I wouldn't worry about the small diff caused by new Response, it still runs at least 2x that of Deno
